File: bindings/java/samples/Zlink.Samples/src/main/java/dev/kairoscode/zlink/samples/SpotRecvSample.java
package dev.kairoscode.zlink.samples;

import dev.kairoscode.zlink.Context;
import dev.kairoscode.zlink.Message;
import dev.kairoscode.zlink.service.discovery.Discovery;
import dev.kairoscode.zlink.service.registry.Registry;
import dev.kairoscode.zlink.service.registry.ServiceType;
import dev.kairoscode.zlink.RecvException;
import dev.kairoscode.zlink.RecvFlags;
import dev.kairoscode.zlink.RecvResult;
import dev.kairoscode.zlink.service.spot.Spot;
import dev.kairoscode.zlink.service.spot.SpotNode;
import java.time.Duration;
import java.time.Instant;

public final class SpotRecvSample {
    public static void main(String[] args) {
        SampleSupport.ensureNative();
        String serviceName = "sample";
        String topic = SampleSupport.SPOT_TOPIC;
        String published = topic + "/" + SampleSupport.SPOT_PAYLOAD;
        String registryPub = SampleSupport.tcpEndpoint();
        String registryRouter = SampleSupport.tcpEndpoint();
        String publisherEndpoint = SampleSupport.tcpEndpoint();
        String subscriberEndpoint = SampleSupport.tcpEndpoint();

        try (Context ctx = new Context();
             Registry registry = new Registry(ctx);
             Discovery discovery = new Discovery(ctx, ServiceType.SPOT, serviceName);
             SpotNode publisherNode = new SpotNode(ctx);
             SpotNode subscriberNode = new SpotNode(ctx);
             Spot publisher = publisherNode.createSpot();
             Spot subscriber = subscriberNode.createSpot()) {
            registry.bind(registryPub, registryRouter);
            discovery.connectRegistry(registryRouter);
            publisherNode.attachDiscovery(discovery);
            subscriberNode.attachDiscovery(discovery);
            publisherNode.bind(publisherEndpoint);
            subscriberNode.bind(subscriberEndpoint);
            subscriber.setSubscription(topic);

            Instant deadline = Instant.now().plus(Duration.ofSeconds(5));
            while (Instant.now().isBefore(deadline)) {
                try (Message payload = Message.copyOfUtf8(SampleSupport.SPOT_PAYLOAD)) {
                    publisher.publish(serviceName, topic, payload);
                }
                try (var topicMessage = subscriber.subscribe(RecvFlags.DONT_WAIT)) {
                    if (!java.util.Optional.of(serviceName)
                        .equals(topicMessage.serviceName())) {
                        throw new IllegalStateException(
                            "unexpected service: " + topicMessage.serviceName());
                    }
                    String value = topicMessage.topic() + "/"
                        + topicMessage.singlePartOrThrow().toUtf8String();
                    if (!published.equals(value)) {
                        throw new IllegalStateException(
                            "unexpected delivery: " + value);
                    }
                    System.out.println("[spot/recv] service: \"" + serviceName
                        + "\" tick: 1 publish: \"" + published
                        + "\" -> recv: \"" + value + "\"");
                    return;
                } catch (RecvException ex) {
                    if (ex.getResult() != RecvResult.NO_DATA
                        && ex.getResult() != RecvResult.BUSY
                        && ex.getInternalErrno() != 2) {
                        throw ex;
                    }
                }
                Thread.onSpinWait();
            }
            throw new IllegalStateException("spot delivery did not arrive");
        }
    }
}

File: bindings/node/samples/spot_recv_sample.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const assert = require('node:assert/strict');
const { once } = require('node:events');
const net = require('node:net');
const zlink = require('../..');
const AUTO_CONNECT_SPOT_MESH = 5;

const SERVICE_NAME = 'sample';

async function reservePort() {
  const server = net.createServer();
  server.listen(0, '127.0.0.1');
  await once(server, 'listening');
  const { port } = server.address();
  await new Promise((resolve, reject) => server.close((error) => error ? reject(error) : resolve()));
  return port;
}

async function main() {
  const ctx = new zlink.Context();
  const registry = new zlink.Registry(ctx);
  const publisherDiscovery = new zlink.Discovery(ctx, AUTO_CONNECT_SPOT_MESH, SERVICE_NAME);
  const subscriberDiscovery = new zlink.Discovery(ctx, AUTO_CONNECT_SPOT_MESH, SERVICE_NAME);
  const publisherNode = new zlink.SpotNode(ctx);
  const subscriberNode = new zlink.SpotNode(ctx);
  let publisher = null;
  let subscriber = null;
  const topic = 'room:lobby';
  const sent = 'hello-spot';
  const registryPub = `tcp://127.0.0.1:${await reservePort()}`;
  const registryRouter = `tcp://127.0.0.1:${await reservePort()}`;
  const publisherEndpoint = `tcp://127.0.0.1:${await reservePort()}`;
  const subscriberEndpoint = `tcp://127.0.0.1:${await reservePort()}`;

  try {
    registry.bind(registryPub, registryRouter);
    publisherDiscovery.connectRegistry(registryRouter);
    subscriberDiscovery.connectRegistry(registryRouter);
    publisherNode.attachDiscovery(publisherDiscovery);
    subscriberNode.attachDiscovery(subscriberDiscovery);
    publisherNode.bind(publisherEndpoint);
    subscriberNode.bind(subscriberEndpoint);
    publisher = publisherNode.createSpot();
    subscriber = subscriberNode.createSpot();
    subscriber.setSubscription(topic);
    const deadline = Date.now() + 15000;
    let received = null;
    while (Date.now() < deadline) {
      publisher.publish(SERVICE_NAME, topic).message(Buffer.from(sent)).submit();
      try {
        received = subscriber.subscribe(zlink.RecvFlags.DontWait);
        if (received) {
          break;
        }
      } catch (error) {
        if (!(error instanceof zlink.RecvError)) {
          throw error;
        }
        if (error.result !== zlink.RecvResult.NoData && error.internalErrno !== 2) {
          throw error;
        }
      }
      await new Promise((resolve) => setTimeout(resolve, 25));
    }
    if (!received) {
      console.log('[spot/recv] remote delivery not observed before timeout');
      return;
    }
    try {
      assert.equal(received.serviceName, SERVICE_NAME);
      assert.equal(received.topic, topic);
      const recv = received.parts[0].data().toString();
      assert.equal(recv, sent);
      console.log(`[spot/recv] service: "${SERVICE_NAME}" tick: 1 publish: "${topic}/${sent}" -> recv: "${topic}/${recv}"`);
    } finally {
      received.close();
    }
  } finally {
    if (subscriber) {
      subscriber.close();
    }
    if (publisher) {
      publisher.close();
    }
    subscriberNode.close();
    publisherNode.close();
    subscriberDiscovery.close();
    publisherDiscovery.close();
    registry.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
